### ISSUE TITLE
Pass arbitrary URL Params on all requests to OIDC Provider

### ISSUE THREAD

**Issue by etiennedi** [jenkinsci/oic-auth-plugin#26]
Thanks for this plugin, I have a small feature-request which I believe we could contribute.

*tl;dr: It would help us tremendously if we could pass arbitrary data as url params to the oidc token issuer. Reason fro that is a custom implementation.*

## How Keycloak does it
The OIDC provider we are using is custom-built at an enterprise company. It provides realms, similar to how keycloak provides realms. However, with keycloak, realms are a top-level namespace, like so:

```
Discovery: https://keycloak.local/auth/realms/master/.well-known/openid-configuration
Authorization: https://keycloak.local/auth/realms/master/protocol/openid-connect/auth
```
If you wanted to change the realm, you'd change it in the path and all URLs (including the discovery) are affected:
```
Discovery: https://keycloak.local/auth/realms/other-realm/.well-known/openid-configuration
Authorization: https://keycloak.local/auth/realms/other-realm/protocol/openid-connect/auth
```

## How our custom solution does it
In our solution unfortunately all URLs only exist once, like so. This implies the "GLOBAL REALM".
```
Discovery: https://custom-oidc.local/.well-known/openid-configuration
Authorization: https://custom-oidc.local/openid-connect/auth
```
If we wanted to change the realm, the path of the urls wouldn't change, but an additional url param would be added, like so:
```
Discovery: https://custom-oidc.local/.well-known/openid-configuration
Authorization: https://custom-oidc.local/openid-connect/auth?realm_id=other-realm
```

## What we need from `oic-auth-plugin`
Since the URL param is not part of the "base url", we cannot specify the realm at all. This currently leads us to seeing an error like this: 
<img width="2962" alt="screen shot 2018-04-06 at 10 16 00" src="https://user-images.githubusercontent.com/8974479/38411611-bc35d2f4-3987-11e8-95f3-c29993791b65.png">

Changing the OIDC Provider is not that easy, unfortunately as this would be a breaking change for many users and the OIDC Provider is outside of our team's control. So for us it would help if we could just pass the realm_id as a url param.

So, we're suggesting an additional field in the config where users can specify arbitrary url params, which will be sent on every request.

## Willingness to contribute

Our team doesn't have any experience writing Jenkins plugins, but this seems like a very minor change. We'd be willing to contribute that, if you believe such an optional field is compatible with your strategy for `oic-auth-plugin`. Please let us know, Thanks.

**Comment by mjmbischoff** [comment]
Ok had a quick look at this:
- Spec saids you MAY which is good.
- Library which we rely on supports it, which makes live easier.

I've put bit of code together. Created a test class etc. to allow you to specify it in an additional configuration field what parameters you always want to pass on. 

I would have to clean up the code a bit, but something I noticed: it seems to already work with the current version if you supply "https://custom-oidc.local/openid-connect/auth?realm_id=other" in the Authorization server url (and probably the other fields as well). I'm mocking things and don't have an oic provider with this behavior handy but I'm pretty confident the mocks are correct.

So now this is usually the point where I check if I haven't misinterpreted the request / specs: where does it currently break?

**Comment by etiennedi** [comment]
Thanks for your quick response. I didn't even consider simply appending the url param to the base. That actually works for the first request against the Authorize endpoint. I do see the login UI from our OIDC Provider. However, because the response type is set to `code`, a second request is made against the Token endpoint after successfully receiving the code. For whatever reason, here the additional url param is ignored, so I'm getting the same error as before ("Client does not have access to realm_id - GLOBAL_REALM").

**Comment by mjmbischoff** [comment]
Just verifying: you added the param to the "Token server url" as well?

**Comment by etiennedi** [comment]
Yes, added the URL param to both fields, but it seems like it's being ignored on the "token" request.

**Comment by mjmbischoff** [comment]
Ok, looking into it further I think it does what you ask but probably not what you want / your implementation expects :-)

`POST /token?foo=bar HTTP/1.1
Accept-Encoding: gzip
User-Agent: Google-HTTP-Java-Client/1.23.0 (gzip)
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
Host: localhost:4444
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Connection: keep-alive
Content-Length: 182

code=code&grant_type=authorization_code&redirect_uri=http%3A%2F%2Flocalhost%3A32901%2Fjenkins%2FsecurityRealm%2FfinishLogin&scope=openid+email&client_id=clientId&client_secret=secret`

As you can see it will send the 'url param' / query string / query param but I strongly expect that your provider expects it in a request body of a POST using Form Serialization as per spec: http://openid.net/specs/openid-connect-core-1_0.html#FormSerialization

Sadly the spec is implicit about if additional parameters are allowed here afaik. It doesn't say it's allowed to pass additional parameters like it explicitly saids elsewhere. It does say the provider is suppose to ignore unknown parameters, which means that correct implementations shouldn't break.

Ok, so a couple of questions:
- can you verify that it's looking at the form encoded parameters?
- do you have control over this custom implementation? 
- since this is starting to feel like a somewhat specific solution, can you perhaps change the impl to also look at the query string?

**Comment by etiennedi** [comment]
> Ok, looking into it further I think it does what you ask but probably not what you want / your implementation expects :-)

I fear, you might be right...

> can you verify that it's looking at the form encoded parameters?
Indeed, I have missed that.

> do you have control over this custom implementation?
Our team is just a consumer of their services, however, the team managing the OIDC provider is part of the same company. 

> since this is starting to feel like a somewhat specific solution, can you perhaps change the impl to also look at the query string?
I agree with you. If it was only the url params it might seem reasonable, but requiring the `realm_id` sometimes in the URL and sometimes in the request body is just arbitrary. I'll hope we can get them to look at the URL params on the `POST /token` as well.

Thanks so much for your input. I'm closing this because - as you said - the feature I requested is already working.
